rectools/models/rank.py

```diff
@@ -0,0 +1,181 @@
+        return -np.finfo(np.float32).max + 1
+
+    @staticmethod
+    def _calc_dots(factors: np.ndarray) -> np.ndarray:
+        return (factors**2).sum(axis=1)
+
+    @staticmethod
+    def _calc_norms(factors: np.ndarray, avoid_zeros: bool = False) -> np.ndarray:
+        norms = np.linalg.norm(factors, axis=1)
+        # Used for COSINE distance
+        # If one or both vectors are zero, assume they're orthogonal, need to avoid 0 in denominator
+        if avoid_zeros:
+            norms[norms == 0] = 1e-10
+        return norms
+
+    def _get_mask_for_correct_scores(self, scores: np.ndarray) -> tp.List[bool]:
+        """Filter scores from implicit library that are not relevant. Implicit library assigns `neginf` score
+        to items that are meant to be filtered (e.g. blacklist items or already seen items)
+        """
+        num_masked = 0
+        min_score = self._get_neginf_score()
+        for el in np.flip(scores):
+            if el <= min_score:
+                num_masked += 1
+            else:
+                break
+        return [True] * (len(scores) - num_masked) + [False] * num_masked
+
+    def _process_implicit_scores(
+        self, subject_ids: InternalIds, ids: np.ndarray, scores: np.ndarray
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+
+        all_target_ids = []
+        all_reco_ids: tp.List[np.ndarray] = []
+        all_scores: tp.List[np.ndarray] = []
+
+        for subject_id, object_ids, object_scores in zip(subject_ids, ids, scores):
+            correct_mask = self._get_mask_for_correct_scores(object_scores)
+            relevant_scores = object_scores[correct_mask]
+            relevant_ids = object_ids[correct_mask]
+
+            if self.distance == Distance.COSINE:
+                subject_norm = self.subjects_norms[subject_id]
+                relevant_scores /= subject_norm
+
+            if self.distance == Distance.EUCLIDEAN:
+                # Restore Euclidean distances from scores
+                d2 = self.subjects_dots[subject_id] - relevant_scores
+                # Theoretically d2 >= 0, but can be <0 because of rounding errors
+                relevant_scores = np.sqrt(np.maximum(d2, 0))
+
+            all_target_ids.extend([subject_id for _ in range(len(relevant_ids))])
+            all_reco_ids.append(relevant_ids)
+            all_scores.append(relevant_scores)
+
+        return all_target_ids, np.concatenate(all_reco_ids), np.concatenate(all_scores)
+
+    def rank(
+        self,
+        subject_ids: InternalIds,
+        k: int,
+        filter_pairs_csr: tp.Optional[sparse.csr_matrix] = None,  # subect-object interactions, relevant for u2i case
+        sorted_object_whitelist: tp.Optional[np.ndarray] = None,
+        num_threads: int = 0,
+    ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
+        """Proceed inference using implicit library matrix factorization topk cpu method"""
```

Comment:
Since we're making this a common method for a few model types, please write a full docstring for it and describe arguments.